First cut of gsettings module

ag-wood opened the pull request:
A module to manage the desktop settings using a python gsettings equivalent.

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+version_added: "1.9"
+notes:
+    - become and become_user must be used to specify the target configuration set.
+requirements: [ "PyGObjects"]
```
What's that?  There's a library called PyGObject (singular).

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+except ImportError:
+    HAS_GIO=False
+else:
+    HAS_GIO=True
```
PEP-8 says you should put spaces around the `=`

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+else:
+    HAS_GIO=True
+
+UID  = 1
```
This doesn't seem to be used anywhere.

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+    'gnome-session', 'mate-session', 'xfce4-session',
+    'cinnamon-session','icewm-session', 'openbox-session' ]
+    
+change_msgs = []
```
Why is this global?  It could very well be a local variable inside `main()`.

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+class Gsettings(object):
+    def __init__(self):
+        def proc_name(pid):
+            '''Return process name of process pid'''
```
Docstrings in Python are usually enclosed in tripple double-quotes: `"""..."""`.  Using single quotes feels weird.

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+            '''Return process name of process pid'''
+            try:
+                for attr in open("/proc/{0}/status".format(pid),'r'):
+                    if 'Name' in attr:
```
Maybe better check `if attr.startswith('Name:'):`?

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+        def proc_environ(pid, varname=""):
+            '''Return process environment variable of process pid'''
+            proc_env = open("/proc/{0}/environ".format(pid),'r')
+            env = proc_env.readline()
```
Better use `.read()` here, environment variables might contain embedded newlines.

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+                if varname != "":
+                    prefix = "{0}=".format(varname)
+                    if env_var.startswith(prefix):
+                        return env_var.replace(prefix, '')
```
Use `env_var[len(prefix):]` instead of `.replace()`.

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+            dbus_process = subprocess.Popen('dbus-launch', stdout=subprocess.PIPE)
+            for env_string in dbus_process.stdout.readlines():
+                env_var = env_string[0:env_string.index('=')]
+                env_value = env_string[env_string.index('=')+1:]
```
Simpler/prettier code:

``` python
env_var, env_value = env_string.split('=', 1)
```

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+                    self.session_pid = procfile
+
+        osuser = pwd.getpwuid(os.getuid()).pw_name
+        if self.session_pid <> -1:
```
`<>` is archaic and deprecated (and altogether removed in Python 3), please use `!=`.

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+            sessionuser = osuser
+
+        # Setup the environment and connect to the bus.
+        if self.session_pid <> -1 and osuser == sessionuser:
```
Why convert UIDs to usernames if all you do with them is compare with each other?  Simpler to keep and compare the UIDs.

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+        ''' Read the value of the specified key '''
+        if ':' in schema:
+            path = schema.split(':')[1]
+            schema = schema.split(':')[0]
```
You could do this on one line:

``` python
schema, path = schema.split(':', 1)
```

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+        if self._cleanup:
+           # Stop the dbus daemon if one was started.
+           time.sleep(1)
+           os.kill(int(self.session_pid), signal.SIGTERM)
```
`__del__` should be avoided if possible. If the new session should always be terminated, better install an [atexit](https://docs.python.org/2/library/atexit.html) handler or something.

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+        schema = gsetting['schema']
+        key = gsetting['key']
+        value = gsetting['value']
+        old_setting = gs.get_value(schema, key)
```
Could this possibly raise an exception, if the schema/key are invalid?  If so, you must try/except and convert that to a `module.fail_json()` call.

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+                new_value = GLib.Variant(setting_type, value)
+                gs.set_value(schema, key, new_value)
+                change_msgs.append('{0}/{1} -> old: {2} ; new: {3}'.format(schema, key, old_value, value))
+    gs = None
```
This is not a very obvious way to trigger the cleanup.  Can you define and call a `.cleanup()` method instead?

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+ENV_DBUS = 'DBUS_SESSION_BUS_ADDRESS'
+SESSION_MANAGERS = [
+    'gnome-session', 'mate-session', 'xfce4-session',
+    'cinnamon-session','icewm-session', 'openbox-session' ]
```
PEP-8: can you please find all the commas that are not followed by a space and add the missing space?

There are [tools](https://pypi.python.org/pypi/flake8) that can check Python style for you.

ag-wood replied:
Yes - sorry, I was a little over-enthusiastic.  Have done the pep-8 check and updated.

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+            env = proc_env.readline()
+            for env_var in env.split('\x00'):
+                if varname != "":
+                    prefix = "{0}=".format(varname)
```
You can hoist the emptiness check and prefix computation outside of the loop.

ag-wood replied:
Sorry - I don't follow what you are saying here...

mgedmin replied:
Sorry for being unclear -- I meant this

``` python
        def proc_environ(pid, varname=""):
            '''Return process environment variable of process pid'''
            if not varname:    # <- the if varname != "" check moved here
                return None
            proc_env = open("/proc/{0}/environ".format(pid),'r')
            env = proc_env.read()   # <- changed as per previous suggestion
            prefix = "{0}=".format(varname)  # <-- moved here
            for env_var in env.split('\x00'):
                if env_var.startswith(prefix):
                    ...
```

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+                change_msgs.append('{0}/{1} -> current: {2} ; proposed new: {3}'.format(schema, key, old_value, value))
+            else:
+                new_value = GLib.Variant(setting_type, value)
+                gs.set_value(schema, key, new_value)
```
Could this raise an exception if you supply an invalid value?

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,225 @@
+        ''' Class cleanup '''
+        if self._cleanup:
+           # Stop the dbus daemon if one was started.
+           time.sleep(1)
```
What is the purpose of this sleep?

ag-wood replied:
The sleep is required to prevent a race condition after setting the value and destroying the process.

ag-wood commented:
@mgedmin, I think I have addressed all of the issues you raised.  Thanks for your assistance, I think it is a much better result than what I originally posted.

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,248 @@
+            UID = 1
+            for ln in open('/proc/{0}/status'.format(pid), 'r'):
+                if ln.startswith('Uid:'):
+                    return int(ln.split()[UID])
```
Uh.  I don't see why you use a named constant here but no such thing in `proc_name()`.

I don't think named constants help, you're just taking the 2nd space-separated field, so `ln.split()[1]` is fine.

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,248 @@
+
+class Gsettings(object):
+    def __init__(self):
+        def proc_name(pid):
```
(This nesting of functions is a bit weird, why not move them to the top level?)

ag-wood replied:
It's a technique I've seen used to hide the implementation of functions that are only used inside the class function.  I'm very happy to move them if it's not the done thing. Still learning lots - thanks!

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,248 @@
+            else:
+                raise ValueError('{0} is not valid in the schema: {1}'.format(key, schema))
+        else:
+            raise ValueError('{0} is not a valid schema'.format(schema))
```
This would flow a bit more naturally without nesting the checks:

``` python
if schema not in ...list_schemas():
    raise ...
settings = ...
if key not in ...list_keys(...):
    raise ...
return settings.get_value(key)
```

mgedmin commented on system/gsettings.py:
```diff
@@ -0,0 +1,248 @@
+        """ Write a value to the specified key """
+        if ':' in schema:
+            path = schema.split(':')[1]
+            schema = schema.split(':')[0]
```
The same sort of refactoring I suggested in get_value() applies here too:

``` python
schema, path = schema.split(':', 1)
```

robynbergeron commented:
Hi @ag-wood -- 

Thanks for submitting this new module to Ansible Extras! This module is now in community review, a process that is open to all Ansible users. In order for this module to be approved, it must gain the following votes:

“works_for_me”: If you have tested the module thoroughly, including testing of all of the module’s options, and if the module works for you, please add “works_for_me” in the comments.

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

“needs_revision”: If the module fails to work for you, or if it doesn’t meet guidelines, please add “needs_revision” in the comments with details about what needs to be fixed.

When a module has both “works_for_me” and “passes_guidelines” tags, we will promote the module for inclusion in Ansible Extras.  At this point, you will be expected to maintain the module by fixing bugs and evaluating pull requests in a timely manner. 

Thanks again for submitting your Ansible module!

robynbergeron commented:
(Also, thanks to @mgedmin for your comments so far, much appreciated!)
